BrowserKit/Sources/WebEngine/Utilities/URLFormatter.swift

```diff
@@ -59,67 +59,76 @@ public class DefaultURLFormatter: URLFormatter {
+            if !escapedEntry.contains(".") {
                 return nil
             }
         }
 
-        guard let url = URL(string: entry) else { return nil }
-
-        // Only allow this URL if it's safe
-        let browsingContext = BrowsingContext(type: .internalNavigation, url: url)
-        if securityManager.canNavigateWith(browsingContext: browsingContext) == .allowed {
-            return URL(string: entry)
-        } else {
-            return nil
-        }
+        return checkBrowsingSafety(with: escapedEntry)
     }
 
     // Handle the entry if it has no scheme
     private func handleNoScheme(with entry: String) -> URL? {
-        // First trim white spaces
+        // First trim white spaces and encode any invalid characters
         let trimmedEntry = entry.trimmingCharacters(in: CharacterSet.whitespacesAndNewlines)
+        guard let escapedEntry = trimmedEntry.addingPercentEncoding(withAllowedCharacters: urlAllowed) else {
+            return nil
+        }
 
         // Make sure there's at least one "." in the host. This means
         // we'll allow single-word searches (e.g., "foo") at the expense
         // of breaking single-word hosts without a scheme
-        if !trimmedEntry.contains(".") || trimmedEntry.contains(" ") { return nil }
+        if !escapedEntry.contains(".") { return nil }
 
         // If entry is a valid floating point number, don't fixup
-        if Double(trimmedEntry) != nil {
+        if Double(escapedEntry) != nil {
             return nil
         }
 
+        // We're going to prepend "http://" only if it's not already present
+        let entryPlusScheme = escapedEntry.hasPrefix("http://") || escapedEntry.hasPrefix("https://") ? escapedEntry : "http://\(escapedEntry)"
+
         // If entry doesn't have a valid ending in Public Suffix List
         // and it's not all digits and dot, stop fix up.
-        if !trimmedEntry.trimmingCharacters(in: CharacterSet(charactersIn: "0123456789.")).isEmpty,
-           let maybeUrl = URL(string: "http://\(trimmedEntry.lowercased())"),
+        if !escapedEntry.trimmingCharacters(in: CharacterSet(charactersIn: "0123456789.")).isEmpty,
+           let maybeUrl = URL(string: entryPlusScheme.lowercased()),
            maybeUrl.publicSuffix == nil {
             return nil
         }
 
-        // Make sure entry only has allowed characters in it
-        guard let escapedURL = trimmedEntry.addingPercentEncoding(withAllowedCharacters: urlAllowed) else {
+        return checkBrowsingSafety(with: entryPlusScheme)
+    }
+
+    private func checkBrowsingSafety(with entry: String) -> URL? {
+        guard let url = URL(url: entry) else { return nil }
+
+        // Only proceed if the URL is correctly formed and there is a host
+        guard let components = URLComponents(url: url, resolvingAgainstBaseURL: false),
+            components?.host != nil else {
             return nil
         }
```

Comment:
Fixing some little issues
```suggestion
        guard let url = URL(string: entry) else { return nil }

        // Only proceed if the URL is correctly formed and there is a host
        guard let components = URLComponents(url: url, resolvingAgainstBaseURL: false),
              components.host != nil
        else {
            return nil
        }
```